### ISSUE TITLE
Oni crashing with Yojson.Json_error("Blank input data"):

### ISSUE THREAD

**Issue by izik1** [onivim/oni2#1766]
os: Arch Linux (i3)
commit: #fa0d3d69 (presumably, 2020-05-14's nightly otherwise), AppImage
steps to repro, as a bash script:
```bash
# backup / remove global store file:
mv ~/.config/oni2/store/global/store.json ~/.config/oni2/store/global/_store.json

# control case (nothing bad should happen here)
oni2 --debug --trace -f 
# equivalently:
# oni2 --debug --trace -f .

# avoid contaminating future testcase
rm ~/.config/oni2/store/global/store.json

# exhibit a:
mkdir -p /tmp/onitest/example-a

 oni2 --debug --trace -f /tmp/onitest/example-a
# there should currently be a window that contains the framebuffer that the terminal last had when the window was created. (see ^1)

# in order to stop oni:
# navigate to terminal that opened oni
<ctrl-C>
killall -SIGKILL Oni2_editor

cat /tmp/onitest/example-a/onivim2-crash.log 
# observe that `/tmp/onitest/example-a/onivim2-crash.log` should contain the following: 
# Yojson.Json_error("Blank input data"):

cat ~/.config/oni2/store/global/store.json
# and that the global store file is indeed empty.

# surely, it will work now, right?
 oni2 --debug --trace -f /tmp/onitest/example-a
# oni should crash without opening any window or logging anything to stdout

cat onivim2-crash.log
# this should also contain: Yojson.Json_error("Blank input data"):
# note that _this_ error log is in `.` whereas the last one was in the opened directory.

# we need to clear the global store again
rm ~/.config/oni2/store/global/store.json

# but this time we should remove `onivim2-crash.log` too
rm onivim2-crash.log

# exhibit b:
mkdir -p /tmp/onitest/example-b/1

# this is required otherwise it'll complain that this directory doesn't exist later
mkdir -p ~/.config/oni2/store/21cdfff0b


echo '{"windowX":2,"windowY":40,"windowWidth":800,"windowHeight":600,"windowMazimized":false}' > ~/.config/oni2/store/21cdfff0b/store.json

echo '{ "workspace": "/tmp/onitest/example-b/1" }' > ~/.config/oni2/store/global/store.json

# oni should open normally. But it doesn't.
oni2 --debug --trace -f /tmp/onitest/example-b/1
# kill it as above.

# note that there shouldn't be any crashlog.
# observe one of the first logs should be:
# [INFO]     +0ms Oni2_editor : Startup: Changing folder to: /tmp/onitest/example-b/1
# let's try changing to that directory
pushd .
cd /tmp/onitest/example-b/1

# Let's try with the path again?
oni2 --debug --trace -f /tmp/onitest/example-b/1
# No such luck, kill it as above.
# [INFO]     +0ms Oni2_editor : Startup: Changing folder to: /tmp/onitest/example-b/1 

# Okay, how about without the path?
oni2 --debug --trace -f
# nooope.
# (kill oni again)
# return from that earlier test
popd
```
^1, this is an oni window, not the terminal used to launch it: ![source 1](https://user-images.githubusercontent.com/21113608/81983999-0e6a7d80-95e9-11ea-98a8-8f369f0319d1.png)

Conclusion: Sorry, I don't know exactly what's going on here, but hopefully this might help with diagnosing/solving the problem?

This is a bit cut-off because some things I thought worked didn't, and something I thought wouldn't did, I'd look at the code to see what's going on, but I can't read ReasonML very well.

**Comment by izik1** [comment]
sidenote while reading over this: `windowMazimized` seems to have a typo in it (should be `windowMaximized`?)

**Comment by bryphe** [comment]
Thank you for the detailed and thorough investigation, @izik1 ! Really appreciate it!

> cat onivim2-crash.log
> # this should also contain: Yojson.Json_error("Blank input data"):

We definitely shouldn't be crashing if we can't load the persistence file... we should be more robust in that case. That's straightforward to fix, I fixed it and added a test in #1767 

I'm concerned about this sequence, still, and not sure this is addressed by the fix:
```
mkdir -p /tmp/onitest/example-a

 oni2 --debug --trace -f /tmp/onitest/example-a
# there should currently be a window that contains the framebuffer that the terminal last had when the window was created. (see ^1)

# in order to stop oni:
# navigate to terminal that opened oni
<ctrl-C>
killall -SIGKILL Oni2_editor
```

It seems like Onivim is hanging when trying to load from that directory (usually if the framebuffer is corrupted like that, it means we got to open the window but never got to rendering, which is interesting.) And it might be the root cause of the file corruption. It's possible it's related to this issue: https://github.com/onivim/oni2/pull/1756 - a potential crash when the size changes, which wasn't in `fa0d3d69` - it could be that we load the default, set the size, hit the crash. 

Would you be up for grabbing a log for that step, @izik1 ? Something like:
```
 oni2 --debug --trace --log-file=example-a.log -f /tmp/onitest/example-a
```
Might give us an idea of where that initial hang is occurring.

Thanks again for your help investigating!

**Comment by izik1** [comment]
Here you go:
[example-a.log](https://github.com/onivim/oni2/files/4632678/example-a.log)

If anything looks fishy about it, I'll try to provide extra information if possible, just not sure what that would be.

**Comment by glennsl** [comment]
Hmm, this is mysterious. I wanted to try to reproduce the underlying issue that caused the blank file to be created in the first place, but I'm unfortunately not able to do so using either the nightly or building from source.

I also can't see how the read and write could conflict as the read is blocking and done during instantiation, and the write needs an actual instance to write, meaning the read must have completed before any write can be attempted. It might be possible if you have multiple instances running, but then it definitely wouldn't fail this predictably.

#1768 should fix the crash at least, which is the more serious problem, but I still worry that the underlying issue might cause persisted settings to be lost.

**Comment by izik1** [comment]
The current nightly (for the 15th, although, once again I don't know how you all count days, I'm in UTC-8, so, it might be the 16th for all of you, and it certainly is UTC) _does_ seem to be working, and without any kind of persisted settings loss at that. As far as I can tell, that particular store only stores the current workspace / version, which, for the record, is `"version":"4d413cfa"`, so maybe it would be impossible to tell if its failing to persist anyway because it might just write over the empty one with the new one, after loading the old one?

**Comment by glennsl** [comment]
If it does, it still wouldn't open the correct workspace. And you could also miss out on an update screen. Neither are huge problems, certainly, but still problems. Most probably wouldn't notice, but it could also leave some with a more or less conscious impression of unpredictable or unreliable behaviour.

But good to hear it doesn't crash anymore at least! Thanks for the feedback @izik1

**Comment by bryphe** [comment]
It sounds like this is fixed now with #1768  👍  Let us know if you're still seeing a crash, and thanks for the great bug report!
